pkg/controller/admissionchecks/multikueue/multikueuecluster.go

```diff
@@ -0,0 +1,345 @@
+	rootContext context.Context
+
+	// For unit testing only. There is now need of creating fully functional remote clients in the unit tests
+	// and creating valid kubeconfig content is not trivial.
+	// The full client creation and usage is validated in the integration and e2e tests.
+	builderOverride clientWithWatchBuilder
+}
+
+var _ manager.Runnable = (*clustersReconciler)(nil)
+var _ reconcile.Reconciler = (*clustersReconciler)(nil)
+
+func (c *clustersReconciler) Start(ctx context.Context) error {
+	c.rootContext = ctx
+	return nil
+}
+
+func (c *clustersReconciler) stopAndRemoveCluster(clusterName string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if rc, found := c.clients[clusterName]; found {
+		rc.watchCancel()
+		delete(c.clients, clusterName)
+	}
+}
+
+func (c *clustersReconciler) setRemoteClientConfig(ctx context.Context, clusterName string, kubeconfig []byte) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	client, found := c.clients[clusterName]
+	if !found {
+		client = newRemoteClient(c.client, c.wlUpdateCh)
+		if c.builderOverride != nil {
+			client.builderOverride = c.builderOverride
+		}
+		c.clients[clusterName] = client
+	}
+
+	if err := client.setConfig(c.rootContext, kubeconfig); err != nil {
+		ctrl.LoggerFrom(ctx).Error(err, "reading local workload")
+		delete(c.clients, clusterName)
+		return err
+	}
+	return nil
+}
+
+func (a *clustersReconciler) controllerFor(acName string) (*remoteClient, bool) {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+
+	c, f := a.clients[acName]
+	return c, f
+}
+
+func (c *clustersReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	cluster := &kueuealpha.MultiKueueCluster{}
+	log := ctrl.LoggerFrom(ctx)
+
+	err := c.client.Get(ctx, req.NamespacedName, cluster)
+	if client.IgnoreNotFound(err) != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err != nil || !cluster.DeletionTimestamp.IsZero() {
+		c.stopAndRemoveCluster(req.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// get the kubeconfig
+	kubeConfig, retry, err := c.getKubeConfig(ctx, &cluster.Spec.KubeConfig)
+	if retry {
+		return reconcile.Result{}, err
+	}
+	if err != nil {
+		log.Error(err, "reading kubeconfig")
+		return reconcile.Result{}, c.updateStatus(ctx, cluster, false, "BadConfig", err.Error())
+	}
+
+	if err := c.setRemoteClientConfig(ctx, cluster.Name, kubeConfig); err != nil {
+		log.Error(err, "setting kubeconfig")
+		return reconcile.Result{}, c.updateStatus(ctx, cluster, false, "ClientConnectionFailed", err.Error())
+	}
+	return reconcile.Result{}, c.updateStatus(ctx, cluster, true, "Active", "Connected")
+}
+
+func (c *clustersReconciler) getKubeConfig(ctx context.Context, ref *kueuealpha.KubeConfig) ([]byte, bool, error) {
+	sec := corev1.Secret{}
+	secretObjKey := types.NamespacedName{
+		Namespace: c.configNamespace,
+		Name:      ref.Location,
+	}
+	err := c.client.Get(ctx, secretObjKey, &sec)
+	if err != nil {
+		return nil, !apierrors.IsNotFound(err), err
```

Comment:
Uhm, this could cause the eternally repeated retring, right?
If so, introducing the timeout mechanism would be better. However, we can work on the mechanism in another PR.

Reply:
It makes sense. The controller-runtime retry mechanism would be safe.